### PR TITLE
[feature] support shuffle read with retry when facing IO error.

### LINKED ISSUE #737: [feature] support shuffle read with retry when facing IO error.
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)
Same in spark 
``` scala
 /**
   * Max number of times we will try IO exceptions (such as connection timeouts) per request.
   * If set to 0, we will not do any retries.
   */
  public int maxIORetries() { return conf.getInt(SPARK_NETWORK_IO_MAXRETRIES_KEY, 3); }
```
https://github.com/apache/spark/blob/2878cd84d5d5edcaa9bbc642c12d8c7f6f009328/common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java#L157
**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Ted-Jiang**
# Which issue does this PR close?

Closes #737 .

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: ballista/core/src/client.rs [hunk 1]
```diff
@@ -117,32 +120,69 @@ impl BallistaClient {
...
-                Ok(Box::pin(FlightDataStream::new(stream, schema)))
+            let request = tonic::Request::new(Ticket {
+                ticket: buf.clone().into(),
+            });
+            let result = self.flight_client.do_get(request).await;
+
+            let res = match result {
+                Ok(res) => res,
+                Err(ref err) => {
+                    // IO related error like connection timeout, reset... will warp with Code::Unknown
```
Comment (Ted-Jiang) [inline ballista/core/src/client.rs:143]: ConnectionReset need retry use `Unknown` type
```
(Execution("FetchFailed(\"10.69.148.14\", \"37258e22-f50e-482b-9c0d-d1e84b83d5e5\", 3, 2, \"Status { code: Unknown, message: \\\"transport error\\\", source: Some(tonic::transport::Error(Transport, hyper::Error(Io, Kind(ConnectionReset)))) }\")"))
```

NotSuchFile use `Internal`
```
DataFusion error: ArrowError(ExternalError("Arrow error: External error: Shuffle fetch partition error from Executor 10.69.147.146 : ff16a886-ca14-4a1b-bdc3-560c6adf699c, map_stage 2, map_partition 18, error desc: Status { code: Internal, message: \"Ballista Error: General(\\\"Failed to open partition file at /home
```

File: ballista/core/src/client.rs [hunk 2]
```diff
@@ -117,32 +120,69 @@ impl BallistaClient {
...
 
-        let request = tonic::Request::new(Ticket { ticket: buf.into() });
-
-        let mut stream = self
-            .flight_client
-            .do_get(request)
-            .await
-            .map_err(|e| BallistaError::GrpcActionError(format!("{e:?}")))?
-            .into_inner();
+        for i in 0..IO_RETRIES_TIMES {
```
Comment (yahoNanJing) [inline ballista/core/src/client.rs:123]: Maybe better to extract the retry logic into a function to get the `FlightData`

File: ballista/core/src/client.rs [hunk 3]
```diff
@@ -43,16 +43,19 @@ use crate::serde::protobuf;
...
+use tonic::{Code, Streaming};
 
 /// Client for interacting with Ballista executors.
 #[derive(Clone)]
 pub struct BallistaClient {
     flight_client: FlightServiceClient<tonic::transport::channel::Channel>,
 }
 
+const IO_RETRIES_TIMES: u8 = 3;
+const IO_RETRY_WAIT_TIME_MS: u64 = 3000;
```
Comment (yahoNanJing) [inline ballista/core/src/client.rs:57]: Better to make these configurations configurable.

Comment (Ted-Jiang) [reply ballista/core/src/client.rs:57]: Will fix this, when support global config.

**Pull request merged**